File: core/src/com/megaman/game/entities/impl/enemies/impl/Met.java
package com.megaman.game.entities.impl.enemies.impl;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.ObjectMap;
import com.megaman.game.ConstKeys;
import com.megaman.game.MegamanGame;
import com.megaman.game.animations.Animation;
import com.megaman.game.animations.AnimationComponent;
import com.megaman.game.assets.TextureAsset;
import com.megaman.game.entities.*;
import com.megaman.game.entities.utils.damage.DamageNegotiation;
import com.megaman.game.entities.utils.damage.Damager;
import com.megaman.game.entities.utils.faceable.Faceable;
import com.megaman.game.entities.utils.faceable.Facing;
import com.megaman.game.entities.impl.enemies.Enemy;
import com.megaman.game.entities.impl.explosions.impl.ChargedShotExplosion;
import com.megaman.game.entities.impl.projectiles.ProjectileFactory;
import com.megaman.game.entities.impl.projectiles.impl.Bullet;
import com.megaman.game.entities.impl.projectiles.impl.ChargedShot;
import com.megaman.game.entities.impl.projectiles.impl.Fireball;
import com.megaman.game.health.HealthVals;
import com.megaman.game.shapes.ShapeComponent;
import com.megaman.game.shapes.ShapeHandle;
import com.megaman.game.shapes.ShapeUtils;
import com.megaman.game.sprites.SpriteComponent;
import com.megaman.game.sprites.SpriteHandle;
import com.megaman.game.updatables.UpdatableComponent;
import com.megaman.game.utils.enums.Position;
import com.megaman.game.utils.objs.Timer;
import com.megaman.game.world.*;
import lombok.Getter;
import lombok.Setter;

import java.util.EnumMap;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

public class Met extends Enemy implements Faceable {

    public static final String RUN_ONLY = "RunOnly";
    public static final String RUNNING_ALLOWED = "RunningAllowed";

    private enum MetBehavior {
        SHIELDING,
        POP_UP,
        RUNNING
    }

    private static final float SHIELDING_DUR = 1.15f;
    private static final float RUNNING_DUR = .5f;
    private static final float POP_UP_DUR = .5f;
    private static final float RUN_VEL = 8f;
    private static final float RUN_IN_WATER_VEL = 3f;
    private static final float GRAVITY_Y = -.15f;
    private static final float BULLET_TRAJ_X = 15f;
    private static final float BULLET_TRAJ_Y = .25f;
    private static final float VEL_CLAMP_X = 8f;
    private static final float VEL_CLAMP_Y = 1.5f;

    private final Map<MetBehavior, Timer> metBehavTimers;
    private final Sprite sprite;

    private MetBehavior behav;
    @Getter
    @Setter
    private Facing facing;
    private String type;
    private boolean runOnly;
    private boolean runningAllowed;

    public Met(MegamanGame game) {
        super(game, BodyType.DYNAMIC);
        metBehavTimers = new EnumMap<>(MetBehavior.class) {{
            put(MetBehavior.SHIELDING, new Timer(SHIELDING_DUR));
            put(MetBehavior.RUNNING, new Timer(RUNNING_DUR));
            put(MetBehavior.POP_UP, new Timer(POP_UP_DUR));
        }};
        type = "";
        defineBody();
        sprite = new Sprite();
        putComponent(spriteComponent());
        putComponent(animationComponent());
    }

    @Override
    public void init(Rectangle bounds, ObjectMap<String, Object> data) {
        setBehav(MetBehavior.SHIELDING);
        Vector2 spawn = ShapeUtils.getBottomCenterPoint(bounds);
        ShapeUtils.setBottomCenterToPoint(body.bounds, spawn);
        runOnly = data.containsKey(RUN_ONLY) && (boolean) data.get(RUN_ONLY);
        type = data.containsKey(ConstKeys.TYPE) ? (String) data.get(ConstKeys.TYPE) : "";
        runningAllowed = !data.containsKey(RUNNING_ALLOWED) || (boolean) data.get(RUNNING_ALLOWED);
        boolean right = data.containsKey(ConstKeys.RIGHT) && (boolean) data.get(ConstKeys.RIGHT);
        setFacing(right ? Facing.RIGHT : Facing.LEFT);
    }

    @Override
    protected Map<Class<? extends Damager>, DamageNegotiation> defineDmgNegs() {
        return new HashMap<>() {{
            put(Bullet.class, new DamageNegotiation(10));
            put(Fireball.class, new DamageNegotiation(15));
            put(ChargedShot.class, new DamageNegotiation(HealthVals.MAX_HEALTH));
            put(ChargedShotExplosion.class, new DamageNegotiation(15));
        }};
    }

    protected void defineBody() {
        body.gravityOn = true;
        body.bounds.setSize(.75f * WorldVals.PPM);
        body.velClamp.set(VEL_CLAMP_X * WorldVals.PPM, VEL_CLAMP_Y * WorldVals.PPM);
        Array<ShapeHandle> h = new Array<>();

        // body fixture
        Fixture bodyFixture = new Fixture(this, FixtureType.BODY,
                new Rectangle().setSize(.75f * WorldVals.PPM));
        h.add(new ShapeHandle(bodyFixture.shape, Color.ORANGE));
        body.add(bodyFixture);

        // feet fixture
        Fixture feetFixture = new Fixture(this, FixtureType.FEET,
                new Rectangle().setSize(.15f * WorldVals.PPM, .2f * WorldVals.PPM));
        feetFixture.offset.y = -.375f * WorldVals.PPM;
        h.add(new ShapeHandle(feetFixture.shape, Color.GREEN));
        body.add(feetFixture);

        // shield fixture
        Fixture shieldFixture = new Fixture(this, FixtureType.SHIELD,
                new Rectangle().setSize(.75f * WorldVals.PPM, .5f * WorldVals.PPM));
        shieldFixture.putUserData(ConstKeys.REFLECT, ConstKeys.UP);
        h.add(new ShapeHandle(shieldFixture.shape, () -> shieldFixture.active ? Color.BLUE : Color.GRAY));
        body.add(shieldFixture);

        // damageable fixture
        Fixture damageableFixture = new Fixture(this, FixtureType.DAMAGEABLE,
                new Rectangle().setSize(.75f * WorldVals.PPM));
        h.add(new ShapeHandle(damageableFixture.shape, () -> damageableFixture.active ? Color.RED : Color.GRAY));
        body.add(damageableFixture);

        // damager fixture
        Fixture damagerFixture = new Fixture(this, FixtureType.DAMAGER,
                new Rectangle().setSize(.75f * WorldVals.PPM));
        body.add(damagerFixture);

        // pre-process
        body.preProcess = delta -> {
            body.gravity.y = is(BodySense.FEET_ON_GROUND) ? 0f : GRAVITY_Y * WorldVals.PPM;
            shieldFixture.active = behav == MetBehavior.SHIELDING;
            damageableFixture.active = behav != MetBehavior.SHIELDING;
        };

        if (MegamanGame.DEBUG) {
            putComponent(new ShapeComponent(h));
        }
    }

    @Override
    protected void defineUpdateComponent(UpdatableComponent c) {
        super.defineUpdateComponent(c);
        c.add(delta -> {
            if (game.getMegaman().dead) {
                return;
            }
            if (runOnly) {
                behav = MetBehavior.RUNNING;
            }
            switch (behav) {
                case SHIELDING -> {
                    Timer shieldingTimer = metBehavTimers.get(MetBehavior.SHIELDING);
                    if (!isPlayerShootingAtMe()) {
                        shieldingTimer.update(delta);
                    }
                    if (shieldingTimer.isFinished()) {
                        setBehav(MetBehavior.POP_UP);
                    }
                }
                case POP_UP -> {
                    setFacing(game.getMegaman().body.isRightOf(body) ? Facing.RIGHT : Facing.LEFT);
                    Timer popupTimer = metBehavTimers.get(MetBehavior.POP_UP);
                    if (popupTimer.isAtBeginning()) {
                        shoot();
                    }
                    popupTimer.update(delta);
                    if (popupTimer.isFinished()) {
                        setBehav(runningAllowed ? MetBehavior.RUNNING : MetBehavior.SHIELDING);
                    }
                }
                case RUNNING -> {
                    Timer runningTimer = metBehavTimers.get(MetBehavior.RUNNING);
                    body.velocity.x = (is(BodySense.BODY_IN_WATER) ? RUN_IN_WATER_VEL : RUN_VEL) * WorldVals.PPM;
                    if (is(Facing.LEFT)) {
                        body.velocity.x *= -1f;
                    }
                    if (!runOnly) {
                        runningTimer.update(delta);
                    }
                    if (runningTimer.isFinished()) {
                        if (is(BodySense.FEET_ON_GROUND)) {
                            body.velocity.x = 0f;
                        }
                        setBehav(MetBehavior.SHIELDING);
                    }
                }
            }
        });
    }

    private void setBehav(MetBehavior behav) {
        this.behav = behav;
        for (Timer timer : metBehavTimers.values()) {
            timer.reset();
        }
    }

    private void shoot() {
        Vector2 traj = new Vector2(BULLET_TRAJ_X, BULLET_TRAJ_Y);
        if (is(Facing.LEFT)) {
            traj.x *= -1f;
        }
        float offset = WorldVals.PPM / 64f;
        Vector2 spawn = new Vector2().set(body.getCenter()).add(is(Facing.LEFT) ? -offset : offset, offset);
        ObjectMap<String, Object> data = new ObjectMap<>();
        data.put(ConstKeys.OWNER, this);
        data.put(ConstKeys.TRAJECTORY, traj);
        Entity e = game.getEntityFactories().fetch(EntityType.PROJECTILE, ProjectileFactory.BULLET);
        game.getGameEngine().spawn(e, spawn, data);
    }

    private SpriteComponent spriteComponent() {
        sprite.setSize(1.5f * WorldVals.PPM, 1.5f * WorldVals.PPM);
        SpriteHandle h = new SpriteHandle(sprite, 4);
        h.updatable = delta -> {
            h.setPosition(body.bounds, Position.BOTTOM_CENTER);
            sprite.setFlip(is(Facing.LEFT), false);
            h.hidden = dmgBlink;
        };
        return new SpriteComponent(h);
    }

    private AnimationComponent animationComponent() {
        Supplier<String> keySupplier = () -> type + switch (behav) {
            case RUNNING -> "Run";
            case POP_UP -> "PopUp";
            case SHIELDING -> "LayDown";
        };
        TextureAtlas atlas = game.getAssMan().getTextureAtlas(TextureAsset.ENEMIES_1);
        return new AnimationComponent(sprite, keySupplier, new ObjectMap<>() {{
            put("Run", new Animation(atlas.findRegion("Met/Run"), 2, .125f));
            put("PopUp", new Animation(atlas.findRegion("Met/PopUp")));
            put("LayDown", new Animation(atlas.findRegion("Met/LayDown")));
            put("SnowRun", new Animation(atlas.findRegion("SnowMet/Run"), 2, .125f));
            put("SnowPopUp", new Animation(atlas.findRegion("SnowMet/PopUp")));
            put("SnowLayDown", new Animation(atlas.findRegion("SnowMet/LayDown")));
        }});
    }

}

File: core/src/main/kotlin/com/megaman/maverick/game/entities/enemies/Met.kt
package com.megaman.maverick.game.entities.enemies

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.utils.Array
import com.mega.game.engine.animations.Animation
import com.mega.game.engine.animations.AnimationsComponent
import com.mega.game.engine.animations.Animator
import com.mega.game.engine.common.GameLogger
import com.mega.game.engine.common.enums.Direction
import com.mega.game.engine.common.enums.Facing
import com.mega.game.engine.common.enums.Position
import com.mega.game.engine.common.enums.Size
import com.mega.game.engine.common.extensions.getTextureAtlas
import com.mega.game.engine.common.extensions.objectMapOf
import com.mega.game.engine.common.interfaces.IDirectional
import com.mega.game.engine.common.interfaces.IFaceable
import com.mega.game.engine.common.objects.Properties
import com.mega.game.engine.common.objects.pairTo
import com.mega.game.engine.common.objects.props
import com.mega.game.engine.common.shapes.GameRectangle
import com.mega.game.engine.common.time.Timer
import com.mega.game.engine.damage.IDamager
import com.mega.game.engine.drawables.shapes.DrawableShapesComponent
import com.mega.game.engine.drawables.shapes.IDrawableShape
import com.mega.game.engine.drawables.sorting.DrawingPriority
import com.mega.game.engine.drawables.sorting.DrawingSection
import com.mega.game.engine.drawables.sprites.GameSprite
import com.mega.game.engine.drawables.sprites.SpritesComponent
import com.mega.game.engine.drawables.sprites.setPosition
import com.mega.game.engine.drawables.sprites.setSize
import com.mega.game.engine.updatables.UpdatablesComponent
import com.mega.game.engine.world.body.Body
import com.mega.game.engine.world.body.BodyComponent
import com.mega.game.engine.world.body.BodyType
import com.mega.game.engine.world.body.Fixture
import com.megaman.maverick.game.ConstKeys
import com.megaman.maverick.game.ConstVals
import com.megaman.maverick.game.MegamanMaverickGame
import com.megaman.maverick.game.assets.SoundAsset
import com.megaman.maverick.game.assets.TextureAsset
import com.megaman.maverick.game.damage.dmgNeg
import com.megaman.maverick.game.entities.MegaEntityFactory
import com.megaman.maverick.game.entities.bosses.GutsTankFist
import com.megaman.maverick.game.entities.contracts.AbstractEnemy
import com.megaman.maverick.game.entities.contracts.megaman
import com.megaman.maverick.game.entities.projectiles.BigAssMaverickRobotOrb
import com.megaman.maverick.game.entities.projectiles.Bullet
import com.megaman.maverick.game.entities.projectiles.PurpleBlast
import com.megaman.maverick.game.utils.GameObjectPools
import com.megaman.maverick.game.utils.extensions.getPositionPoint
import com.megaman.maverick.game.world.body.*

class Met(game: MegamanMaverickGame) : AbstractEnemy(game, size = Size.SMALL), IFaceable, IDirectional {

    enum class MetBehavior { SHIELDING, POP_UP, RUNNING }

    companion object {
        const val TAG = "Met"

        const val RUN_ONLY = "run_only"
        const val RUNNING_ALLOWED = "running_allowed"

        private var atlas: TextureAtlas? = null

        private const val SHIELDING_DURATION = 1.15f
        private const val RUNNING_DURATION = 0.5f
        private const val POP_UP_DURATION = 0.5f

        private const val RUN_SPEED = 8f

        private const val GRAVITY_IN_AIR = 0.25f
        private const val GRAVITY_ON_GROUND = 0.01f

        private const val BULLET_TRAJECTORY_X = 15f
        private const val BULLET_TRAJECTORY_Y = 0.25f

        private const val VELOCITY_CLAMP_X = 8f
        private const val VELOCITY_CLAMP_Y = 18f
    }

    override lateinit var facing: Facing
    override var direction: Direction
        get() = body.direction
        set(value) {
            body.direction = value
        }

    private val metBehaviorTimers = objectMapOf(
        MetBehavior.SHIELDING pairTo Timer(SHIELDING_DURATION),
        MetBehavior.POP_UP pairTo Timer(POP_UP_DURATION),
        MetBehavior.RUNNING pairTo Timer(RUNNING_DURATION)
    )
    private lateinit var type: String
    private var behavior: MetBehavior = MetBehavior.SHIELDING
        set(value) {
            field = value
            metBehaviorTimers.values().forEach { it.reset() }
        }
    private var runOnly = false
    private var runningAllowed = false
    private var runSpeed = RUN_SPEED

    override fun init() {
        GameLogger.debug(TAG, "init()")
        super.init()
        if (atlas == null) atlas = game.assMan.getTextureAtlas(TextureAsset.ENEMIES_1.source)
        addComponent(defineAnimationsComponent())
        damageOverrides.put(GutsTankFist::class, dmgNeg(ConstVals.MAX_HEALTH))
        damageOverrides.put(BigAssMaverickRobotOrb::class, dmgNeg(ConstVals.MAX_HEALTH))
        damageOverrides.put(PurpleBlast::class, dmgNeg(ConstVals.MAX_HEALTH))
    }

    override fun onSpawn(spawnProps: Properties) {
        GameLogger.debug(TAG, "onSpawn(): spawnProps=$spawnProps")
        super.onSpawn(spawnProps)

        behavior = MetBehavior.SHIELDING

        val spawn = if (spawnProps.containsKey(ConstKeys.BOUNDS)) {
            val bounds = spawnProps.get(ConstKeys.BOUNDS) as GameRectangle
            bounds.getPositionPoint(Position.BOTTOM_CENTER)
        } else spawnProps.get(ConstKeys.POSITION, Vector2::class)!!
        body.setBottomCenterToPoint(spawn)

        runningAllowed = spawnProps.getOrDefault(RUNNING_ALLOWED, true, Boolean::class)
        runOnly = spawnProps.getOrDefault(RUN_ONLY, false, Boolean::class)
        runSpeed = spawnProps.getOrDefault(ConstKeys.SPEED, RUN_SPEED, Float::class)

        type = spawnProps.getOrDefault(ConstKeys.TYPE, "", String::class)

        val right = spawnProps.getOrDefault(ConstKeys.RIGHT, megaman.body.getX() > body.getX(), Boolean::class)
        facing = if (right) Facing.RIGHT else Facing.LEFT

        direction = Direction.UP
    }

    override fun onDestroy() {
        GameLogger.debug(TAG, "onDestroy()")
        super.onDestroy()
    }

    override fun canBeDamagedBy(damager: IDamager) =
        damageOverrides.containsKey(damager::class) || super.canBeDamagedBy(damager)

    override fun takeDamageFrom(damager: IDamager): Boolean {
        if (damageOverrides.containsKey(damager::class)) explode()
        return super.takeDamageFrom(damager)
    }

    private fun shoot() {
        GameLogger.debug(TAG, "shoot()")

        val trajectory = (when (direction) {
            Direction.UP -> Vector2(BULLET_TRAJECTORY_X * facing.value, BULLET_TRAJECTORY_Y)
            Direction.DOWN -> Vector2(BULLET_TRAJECTORY_X * facing.value, -BULLET_TRAJECTORY_Y)
            Direction.LEFT -> Vector2(BULLET_TRAJECTORY_Y, BULLET_TRAJECTORY_X * facing.value)
            Direction.RIGHT -> Vector2(BULLET_TRAJECTORY_Y, -BULLET_TRAJECTORY_X * facing.value)
        }).scl(ConstVals.PPM.toFloat())

        val offset = 0.1f * ConstVals.PPM
        val spawn = body.getCenter().add(offset * facing.value, if (direction == Direction.DOWN) offset else -offset)

        val spawnProps = props(
            ConstKeys.OWNER pairTo this, ConstKeys.TRAJECTORY pairTo trajectory, ConstKeys.POSITION pairTo spawn
        )
        val bullet = MegaEntityFactory.fetch(Bullet::class)!!
        bullet.spawn(spawnProps)

        requestToPlaySound(SoundAsset.ENEMY_BULLET_SOUND, false)
    }

    override fun defineUpdatablesComponent(updatablesComponent: UpdatablesComponent) {
        super.defineUpdatablesComponent(updatablesComponent)
        updatablesComponent.add {
            if (runOnly) behavior = MetBehavior.RUNNING

            when (behavior) {
                MetBehavior.SHIELDING -> {
                    when (direction) {
                        Direction.UP, Direction.DOWN -> body.physics.velocity.x = 0f
                        Direction.LEFT, Direction.RIGHT -> body.physics.velocity.y = 0f
                    }
                    val shieldTimer = metBehaviorTimers.get(MetBehavior.SHIELDING)
                    if (!isMegamanShootingAtMe() && body.isSensing(BodySense.FEET_ON_GROUND)) shieldTimer.update(it)
                    if (shieldTimer.isFinished()) behavior = MetBehavior.POP_UP
                }

                MetBehavior.POP_UP -> {
                    if (!body.isSensing(BodySense.FEET_ON_GROUND)) {
                        behavior = MetBehavior.SHIELDING
                        return@add
                    }

                    facing = when (direction) {
                        Direction.UP, Direction.DOWN -> if (megaman.body.getX() > body.getX()) Facing.RIGHT else Facing.LEFT
                        Direction.LEFT, Direction.RIGHT -> if (megaman.body.getY() > body.getY()) Facing.RIGHT else Facing.LEFT
                    }

                    val popUpTimer = metBehaviorTimers.get(MetBehavior.POP_UP)
                    if (popUpTimer.isAtBeginning()) shoot()
                    popUpTimer.update(it)
                    if (popUpTimer.isFinished()) behavior =
                        if (runningAllowed) MetBehavior.RUNNING else MetBehavior.SHIELDING
                }

                MetBehavior.RUNNING -> {
                    if (!body.isSensing(BodySense.FEET_ON_GROUND)) {
                        behavior = MetBehavior.SHIELDING
                        return@add
                    }

                    val runningTimer = metBehaviorTimers.get(MetBehavior.RUNNING)

                    val runImpulse =
                        ConstVals.PPM * facing.value * if (body.isSensing(BodySense.IN_WATER)) (runSpeed / 2f) else runSpeed
                    when (direction) {
                        Direction.UP, Direction.DOWN -> body.physics.velocity.x = runImpulse
                        Direction.LEFT, Direction.RIGHT -> body.physics.velocity.y = runImpulse
                    }

                    if (!runOnly) {
                        runningTimer.update(it)
                        if (runningTimer.isFinished()) {
                            if (body.isSensing(BodySense.FEET_ON_GROUND)) body.physics.velocity.setZero()
                            behavior = MetBehavior.SHIELDING
                        }
                    }
                }
            }
        }
    }

    override fun defineBodyComponent(): BodyComponent {
        val body = Body(BodyType.DYNAMIC)
        body.physics.applyFrictionX = false
        body.physics.applyFrictionY = false
        body.setSize(ConstVals.PPM.toFloat())

        val debugShapes = Array<() -> IDrawableShape?>()
        debugShapes.add { body.getBounds() }

        val feetFixture =
            Fixture(body, FixtureType.FEET, GameRectangle().setSize(ConstVals.PPM.toFloat(), 0.1f * ConstVals.PPM))
        feetFixture.offsetFromBodyAttachment.y = -body.getHeight() / 2f
        body.addFixture(feetFixture)
        feetFixture.drawingColor = Color.GREEN
        debugShapes.add { feetFixture }

        body.preProcess.put(ConstKeys.DEFAULT) {
            body.physics.velocityClamp.set(
                when {
                    direction.isVertical() -> Vector2(VELOCITY_CLAMP_X, VELOCITY_CLAMP_Y)
                    else -> Vector2(VELOCITY_CLAMP_Y, VELOCITY_CLAMP_X)
                }
            ).scl(ConstVals.PPM.toFloat())

            val gravity = if (body.isSensing(BodySense.FEET_ON_GROUND)) GRAVITY_ON_GROUND else GRAVITY_IN_AIR
            val gravityVec = GameObjectPools.fetch(Vector2::class)
            when (direction) {
                Direction.UP -> gravityVec.set(0f, -gravity)
                Direction.DOWN -> gravityVec.set(0f, gravity)
                Direction.LEFT -> gravityVec.set(gravity, 0f)
                Direction.RIGHT -> gravityVec.set(-gravity, 0f)
            }.scl(ConstVals.PPM.toFloat())
            body.physics.gravity.set(gravityVec)

            body.fixtures.get(FixtureType.SHIELD).first().let {
                it.setActive(behavior == MetBehavior.SHIELDING)
                it.putProperty(ConstKeys.DIRECTION, direction)
            }
            body.fixtures.get(FixtureType.DAMAGEABLE).first().setActive(behavior != MetBehavior.SHIELDING)
        }

        addComponent(DrawableShapesComponent(debugShapeSuppliers = debugShapes, debug = true))

        return BodyComponentCreator.create(
            this,
            body,
            BodyFixtureDef.of(FixtureType.BODY, FixtureType.DAMAGER, FixtureType.DAMAGEABLE, FixtureType.SHIELD)
        )
    }

    override fun defineSpritesComponent(): SpritesComponent {
        val sprite = GameSprite(DrawingPriority(DrawingSection.PLAYGROUND, 4))
        sprite.setSize(2f * ConstVals.PPM)
        val component = SpritesComponent(sprite)
        component.putUpdateFunction { _, _ ->
            sprite.hidden = damageBlink

            val flipX = facing == Facing.LEFT
            val flipY = direction == Direction.DOWN
            sprite.setFlip(flipX, flipY)

            val rotation = when (direction) {
                Direction.UP, Direction.DOWN -> 0f
                Direction.LEFT -> 90f
                Direction.RIGHT -> 270f
            }
            sprite.setOriginCenter()
            sprite.rotation = rotation

            val position = when (direction) {
                Direction.UP -> Position.BOTTOM_CENTER
                Direction.DOWN -> Position.TOP_CENTER
                Direction.LEFT -> Position.CENTER_RIGHT
                Direction.RIGHT -> Position.CENTER_LEFT
            }
            val bodyPosition = body.getPositionPoint(position)
            sprite.setPosition(bodyPosition, position)
        }

        return component
    }

    private fun defineAnimationsComponent(): AnimationsComponent {
        val keySupplier: (String?) -> String? = {
            "$type${
                when (behavior) {
                    MetBehavior.SHIELDING -> "LayDown"
                    MetBehavior.POP_UP -> "PopUp"
                    MetBehavior.RUNNING -> "Run"
                }
            }"
        }
        val animator = Animator(
            keySupplier, objectMapOf(
                "Run" pairTo Animation(atlas!!.findRegion("Met/Run"), 1, 2, 0.125f, true),
                "PopUp" pairTo Animation(atlas!!.findRegion("Met/PopUp"), false),
                "LayDown" pairTo Animation(atlas!!.findRegion("Met/LayDown"), false),
                "SnowRun" pairTo Animation(atlas!!.findRegion("SnowMet/Run"), 1, 2, 0.125f, true),
                "SnowPopUp" pairTo Animation(atlas!!.findRegion("SnowMet/PopUp"), false),
                "SnowLayDown" pairTo Animation(atlas!!.findRegion("SnowMet/LayDown"), false)
            )
        )
        return AnimationsComponent(this, animator)
    }

    override fun getTag() = TAG
}
